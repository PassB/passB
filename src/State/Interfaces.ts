export type ExtensionName = string;
export type StrategyType = 'FileFormat' | 'Filler' | 'Matcher';
export type StrategyName = string;
export type ActionName = string;
