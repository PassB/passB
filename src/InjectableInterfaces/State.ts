import {Store} from 'redux';
import {Persistor} from 'redux-persist';
import {StoreContents} from '../State/State';

export interface State {
  hydrated: Promise<void>;

  getPersistor(): Persistor;

  getStore(): Store<StoreContents>;
}
