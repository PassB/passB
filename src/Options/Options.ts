import {PassExtension} from "Extensions/PassExtension/PassExtension";
import {PassB} from "PassB";
import {OptionsReceiverInterface} from "./OptionsReceiver";

import deepExtend = require("deep-extend");

interface OptionsList {
  [extensionName: string]: {};
}

export interface OptionsData {
  enabledExtensions: string[];
  extensionsOptions: OptionsList;
  selectedMatcher: string;
  matchers: OptionsList;
  selectedFileFormat: string;
  fileFormats: OptionsList;
  selectedFiller: string;
  fillers: OptionsList;
}

function extractDefaultOptionsList(receivers: Array<OptionsReceiverInterface<{}>>): OptionsList {
  return receivers.reduce((list: OptionsList, receiver: OptionsReceiverInterface<{}>): OptionsList => ({
    ...list,
    [receiver.name]: receiver.defaultOptions,
  }), {} as OptionsList);
}

export class Options {
  private _options: OptionsData;

  public constructor(private passB: PassB) {
  }

  public async getOptions(): Promise<OptionsData> {
    if (!this._options) {
      this._options = deepExtend(
        this.getDefaultOptions(this.passB),
        (await browser.storage.local.get('options')) as OptionsData,
      );
    }
    return deepExtend({}, this._options);
  }

  public setOptions(newOptions: OptionsData): Promise<void> {
    this._options = newOptions;
    return browser.storage.local.set({options: newOptions});
  }

  private getDefaultOptions(passB: PassB): OptionsData {
    return {
      enabledExtensions: [PassExtension.name],
      extensionsOptions: extractDefaultOptionsList(passB.getAllExtensions()),
      selectedMatcher: passB.getAllMatchers()[0].constructor.name,
      matchers: extractDefaultOptionsList(passB.getAllMatchers()),
      selectedFileFormat: passB.getAllFileFormats()[0].constructor.name,
      fileFormats: extractDefaultOptionsList(passB.getAllFileFormats()),
      selectedFiller: passB.getAllFillers()[0].constructor.name,
      fillers: extractDefaultOptionsList(passB.getAllFillers()),
    };
  }
}
