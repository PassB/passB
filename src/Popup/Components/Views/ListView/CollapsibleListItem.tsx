import {ListItem, ListItemText} from 'material-ui';
import {ExpandLess, ExpandMore, Folder, InsertDriveFile} from 'material-ui-icons';
import {withStyles, ClassProps, Theme} from 'material-ui/styles';
import Collapse from 'material-ui/transitions/Collapse';
import Avatar from 'material-ui/Avatar';
import * as React from 'react';
import {EntryNode} from 'PassB';
import {EntryNodeList} from './EntryNodeList';

interface Props {
  node: EntryNode;
}

interface State {
  collapsed: boolean;
}

const styles = (theme: Theme) => ({
  nested: {
    paddingLeft: theme.spacing.unit * 4,
  },
});

export const CollapsibleListItem = withStyles(styles)(
  class extends React.Component<Props & ClassProps<{ nested: Partial<React.CSSProperties> }>, State> {
    public state: State = {
      collapsed: true,
    };

    public render(): JSX.Element {
      const {node, classes} = this.props;
      const {collapsed} = this.state;

      return (
        <div>
          <ListItem
            button={true}
            key={node.fullPath}
            onClick={() => this.setState({collapsed: !collapsed})}
          >
            <Avatar>
              {node.fullPath.endsWith('/') ?
                <Folder/> :
                <InsertDriveFile/>
              }
            </Avatar>
            <ListItemText primary={node.fullPath.replace(/\//g, '/\u200b')}/>
            {collapsed ?
              <ExpandMore/> :
              <ExpandLess/>
            }
          </ListItem>
          <Collapse in={!collapsed} transitionDuration="auto" unmountOnExit={true}>
            <div className={classes.nested}>
              <EntryNodeList root={node}/>
            </div>
          </Collapse>
        </div>
      );
    }
  });
