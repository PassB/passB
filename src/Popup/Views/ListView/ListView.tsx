import {passB} from "ConfiguredPassB";
import Menu from 'material-ui/Menu';
import MenuItem from 'material-ui/MenuItem';
import {Entry, LabeledEntries} from "PassB";
import * as React from 'react';
import Tab = browser.tabs.Tab;

interface Props {
  url: string;
  navigateTo: (newUrl: string, state: {}) => void;
}

interface State {
  entries: Entry[];
  filtered: Entry[];
}

export class ListView extends React.Component<Props, State> {
  public state: State = {
    entries: [],
    filtered: [],
  };

  public componentDidMount(): void {
    passB.getEntries()
      .then((entries: LabeledEntries) =>
        this.setState({entries: Object.values(entries)}, () => this.recalculateFilteredEntries()),
      );
  }

  public render(): JSX.Element {
    const {navigateTo} = this.props;
    const {filtered} = this.state;

    return (
      <Menu autoWidth={false} width={400} maxHeight={400}>
        {filtered && filtered.map((entry: Entry) => (
          <MenuItem
            key={entry.label}
            onClick={() => navigateTo(`entry`, {entry})}
            primaryText={entry.label.replace(/\//g, "/\u200b")}
          />
        ))}
      </Menu>
    );
  }

  private recalculateFilteredEntries(): void {
    if (!this.props.url || !this.state.entries) {
      return;
    }
    passB.getMatcher()
      .filterEntries(this.props.url || '', this.state.entries)
      .then((filtered: Entry[]) => this.setState({filtered}));
  }
}
