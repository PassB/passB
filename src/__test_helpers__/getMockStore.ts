import {combineReducers, createStore, Store} from 'redux';
import {StoreContents} from 'InjectableInterfaces/State';
import {initialState as initialOptionsState, reducer as OptionsReducer} from 'State/Options/Reducers';
import {initialState as initialPassEntriesState, reducer as PassEntryReducer} from 'State/PassEntries/Reducers';

const reducer = combineReducers<StoreContents>({
  options: OptionsReducer,
  passEntries: PassEntryReducer,
});

export const getMockStore = (): Store<StoreContents> => createStore<StoreContents>(
  reducer,
  {
    options: initialOptionsState,
    passEntries: initialPassEntriesState,
  },
);
