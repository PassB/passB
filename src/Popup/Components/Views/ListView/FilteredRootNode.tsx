import {ListItemText} from 'material-ui';
import * as React from 'react';
import {connect} from 'react-redux';
import {createSelector} from 'reselect';
import {StoreContents} from 'InjectableInterfaces/State';
import {EntryNode} from 'State/PassEntries/Interfaces';
import {getCombinedPassEntries} from 'State/PassEntries/Selectors';
import {deepFilterEntryNodes, flattenEntryNode} from './entryNodeMethods';
import {CollapsibleListItem} from './CollapsibleListItem';
import {EntryNodeList} from './EntryNodeList';

interface FilteredProps {
  filter: string;
}

export const filteredRootNodeSelector = createSelector(
  getCombinedPassEntries,
  (_: {}, props: FilteredProps) => ({filter: props.filter}),
  (rootNode: EntryNode, {filter}: FilteredProps): EntryNode | undefined => {
    if (!filter) {
      return rootNode;
    }
    return deepFilterEntryNodes(
      rootNode,
      flattenEntryNode(rootNode)
        .filter((node: EntryNode) => node.get('fullPath').includes(filter)),
    );
  },
);

export const FilteredRootNodeComponent = ({root, filter}: { root: EntryNode; filter: string }) => (
  <CollapsibleListItem
    className="allEntriesRoot"
    CollapsedChildren={<EntryNodeList root={root}/>}
    initiallyExpanded={!!filter}
  >
    <ListItemText
      primary={browser.i18n.getMessage('popup_node_all_items')}
    />
  </CollapsibleListItem>
);

export const mapStateToProps = (state: StoreContents, props: FilteredProps) => ({
  root: filteredRootNodeSelector(state, props),
});

export const FilteredRootNode = connect(mapStateToProps)(FilteredRootNodeComponent);
