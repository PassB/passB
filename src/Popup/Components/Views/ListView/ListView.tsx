import {OrderedMap} from 'immutable';
import debounce = require('lodash.debounce');
import {List, ListItem, ListItemText, TextField} from 'material-ui';
import {Sync} from 'material-ui-icons';
import {withStyles, WithStyles} from 'material-ui/styles';
import * as React from 'react';
import {connect} from 'react-redux';
import {compose} from 'recompose';
import {LazyInject} from 'Decorators/LazyInject';
import {PassB} from 'PassB';
import {EntryNode} from 'State/PassEntries/Interfaces';
import {getCombinedPassEntries} from 'State/PassEntries/Selectors';
import {StoreContents} from 'State/State';
import {CollapsibleListItem} from './CollapsibleListItem';
import {EntryNodeList} from './EntryNodeList';

interface Props {
  url: string;
}

interface ConnectedProps {
  rootNode: EntryNode;
}

interface State {
  filteredRootNode?: EntryNode;
  contextualRootNode?: EntryNode;
}

const styles = {
  centered: {
    width: '100%',
    textAlign: 'center',
  },
  noTopPadding: {
    paddingTop: '0px',
  },
};

function flattenEntryNode(currentNode: EntryNode, flattened: EntryNode[] = []): EntryNode[] {
  flattened.push(currentNode);
  currentNode.get('children').forEach((item: EntryNode) => flattenEntryNode(item, flattened));

  return flattened;
}

function deepFilterEntryNodes(node: EntryNode, filteredNodes: EntryNode[]): EntryNode {
  const nodeFullNames = filteredNodes.map((filtered: EntryNode) => filtered.get('fullPath'));
  return node.update(
    'children',
    (origChildren: OrderedMap<string, EntryNode>) =>
      origChildren
        .map((child: EntryNode) => deepFilterEntryNodes(child, filteredNodes))
        .filter((child: EntryNode) => nodeFullNames.includes(child.get('fullPath')) || child.get('children').count() > 0),
  );
}

type InnerProps = Props & ConnectedProps & WithStyles<keyof typeof styles>;

class ListViewComponent extends React.Component<InnerProps, State> {
  public state: State = {};

  @LazyInject(() => PassB)
  private passB: PassB;

  public constructor(props: InnerProps) {
    super(props);

    this.filterNodes = debounce(this.filterNodes.bind(this), 50);
  }

  public componentDidMount(): void {
    this.recalculateContextualEntries(this.props);
  }

  public componentWillReceiveProps(newProps: InnerProps): void {
    if (newProps.url !== this.props.url || newProps.rootNode !== this.props.rootNode) {
      this.recalculateContextualEntries(newProps);
    }
  }

  public render(): JSX.Element {
    const {classes, rootNode} = this.props;
    const {contextualRootNode, filteredRootNode} = this.state;

    return (
      <List>
        <ListItem className={classes.noTopPadding}>
          <TextField
            id="search"
            label={browser.i18n.getMessage('popup_search')}
            type="search"
            fullWidth={true}
            margin="none"
            onChange={(e: React.ChangeEvent<HTMLInputElement>) => this.filterNodes(e.target.value)}
            inputProps={{autoFocus: true, tabIndex: 0}}
          />
        </ListItem>
        {rootNode ?
          <CollapsibleListItem
            key="allEntriesRoot"
            CollapsedChildren={() => <EntryNodeList root={filteredRootNode || rootNode}/>}
            initiallyExpanded={typeof filteredRootNode !== 'undefined'}
          >
            <ListItemText
              primary="All Items"
            />
          </CollapsibleListItem> :
          <div className={classes.centered}><Sync/></div>
        }
        {contextualRootNode &&
        <CollapsibleListItem
          key="contextualRoot"
          CollapsedChildren={() => <EntryNodeList root={contextualRootNode}/>}
          initiallyExpanded={true}
        >
          <ListItemText
            primary="Contextual"
          />
        </CollapsibleListItem>
        }
      </List>
    );
  }

  private filterNodes(filter: string): void {
    const {rootNode} = this.props;
    const flattened: EntryNode[] = flattenEntryNode(rootNode);
    const filteredNodes = flattened.filter((node: EntryNode) => node.get('fullPath').includes(filter));

    this.setState({filteredRootNode: deepFilterEntryNodes(rootNode, filteredNodes)});

  }

  private async recalculateContextualEntries(props: InnerProps): Promise<void> {
    const {rootNode, url} = props;
    if (!url || !rootNode) {
      return;
    }

    const matcher = await this.passB.getMatcher();
    const filteredNodes = await matcher.filterEntries(url, flattenEntryNode(rootNode));
    this.setState({contextualRootNode: deepFilterEntryNodes(rootNode, filteredNodes)});
  }
}

export const ListView = compose<InnerProps, Props>(
  withStyles<keyof typeof styles>(styles),
  connect<ConnectedProps>((state: StoreContents) => ({
    rootNode: getCombinedPassEntries(state),
  })),
)(ListViewComponent);
