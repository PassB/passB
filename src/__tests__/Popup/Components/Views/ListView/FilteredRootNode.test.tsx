import {shallow} from 'enzyme';
import * as React from 'react';
import {Store} from 'redux';
import {getMockStore} from '__test_helpers__/getMockStore';
import {CollapsibleListItem} from 'Popup/Components/Views/ListView/CollapsibleListItem';
import {mapStateToProps, FilteredRootNodeComponent} from 'Popup/Components/Views/ListView/FilteredRootNode';
import {enableExtension} from 'State/Options/Actions';
import {reducer} from 'State/PassEntries';
import {setEntries} from 'State/PassEntries/Actions';
import {EntryAction, EntryNode} from 'State/PassEntries/Interfaces';
import {StoreContents} from 'State/State';

function mockRoot(entries: EntryAction[]): EntryNode {
  return reducer(undefined!, setEntries({
    extensionName: 'testExtension',
    entries,
  })).get('testExtension')!;
}

describe('FilteredRootNode', () => {
  describe('FilteredRootNodeComponent', () => {
    it('renders CollapsibleListItem that contains an EntryNodeList with root node - initially collapsed', () => {
      const rootNode = mockRoot([]);
      const rendered = shallow(<FilteredRootNodeComponent root={rootNode} filter=""/>);
      expect(rendered.find(CollapsibleListItem).length).toBe(1);
      expect((rendered.find(CollapsibleListItem).prop('CollapsedChildren') as JSX.Element).props.root).toEqualImmutable(rootNode);
      expect(rendered.find(CollapsibleListItem).prop('initiallyExpanded')).toBe(false);
    });

    it('renders with initiallyExpanded when filter is not empty', () => {
      const rootNode = mockRoot([]);
      const rendered = shallow(<FilteredRootNodeComponent root={rootNode} filter="test"/>);
      expect(rendered.find(CollapsibleListItem).prop('initiallyExpanded')).toBe(true);
    });
  });

  describe('mapStateToProps', () => {
    let store: Store<StoreContents>;
    beforeAll(() => {
      store = getMockStore();
      store.dispatch(enableExtension({extensionName: 'testExtension'}));
      store.dispatch(setEntries({
        extensionName: 'testExtension',
        entries: [
          {fullPath: 'example.com/login1/testUser', action: 'test'},
          {fullPath: 'example.com/logout1/testUser', action: 'test'},
          {fullPath: 'example.com/logout2/testUser', action: 'test'},
        ],
      }));
    });

    it('filters for all nodes where fullPath matches partially', () => {
      const result = mapStateToProps(store.getState(), {filter: 'logout'});
      expect(result).toMatchSnapshot();
    });
  });
});
