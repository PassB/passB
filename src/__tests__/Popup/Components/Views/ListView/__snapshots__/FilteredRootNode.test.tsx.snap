// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteredRootNode mapStateToProps filters for all nodes where fullPath matches partially 1`] = `
Object {
  "root": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {
      "example.com/": Immutable.Map {
        "name": "example.com/",
        "fullPath": "example.com/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "logout1/": Immutable.Map {
            "name": "logout1/",
            "fullPath": "example.com/logout1/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "testUser": Immutable.Map {
                "name": "testUser",
                "fullPath": "example.com/logout1/testUser",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
          "logout2/": Immutable.Map {
            "name": "logout2/",
            "fullPath": "example.com/logout2/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "testUser": Immutable.Map {
                "name": "testUser",
                "fullPath": "example.com/logout2/testUser",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
        },
      },
    },
  },
}
`;
